huxley/www/js/stores/DelegateStore.js

```diff
@@ -1,32 +1,94 @@
 /**
- * Copyright (c) 2011-2015 Berkeley Model United Nations. All rights reserved.
+ * Copyright (c) 2011-2016 Berkeley Model United Nations. All rights reserved.
  * Use of this source code is governed by a BSD License (see LICENSE).
  */
 
 'use strict';
 
-var $ = require('jquery');
+var ActionConstants = require('constants/ActionConstants');
+var DelegateActions = require('actions/DelegateActions');
 var Dispatcher = require('dispatcher/Dispatcher');
 var ServerAPI = require('lib/ServerAPI');
 var {Store} = require('flux/utils');
 
 
-var _delegatePromises = {};
+var _schoolsDelegates = {};
+var _delegates = {};
 
 class DelegateStore extends Store {
-  getDelegates(schoolID, callback) {
-    if (!_delegatePromises[schoolID]) {
-      _delegatePromises[schoolID] = ServerAPI.getDelegates(schoolID);
+  getDelegates(schoolID) {
+    if (_schoolsDelegates[schoolID]) {
+      return _schoolsDelegates[schoolID];
     }
-    if (callback) {
-      _delegatePromises[schoolID].then(callback);
+
+    ServerAPI.getDelegates(schoolID).then(value => {
+      _schoolsDelegates[schoolID] = value;
+      for (var i = 0; i < value.length; i++) {
+        var delegate = value[i];
+        _delegates[delegate.id] = delegate;
+      }
+      DelegateActions.delegatesFetched();
+    });
+
+    return [];
+  }
+
+  deleteDelegate(delegateID) {
+    ServerAPI.deleteDelegate(delegateID);
+    var schoolID = _delegates[delegateID].school;
+    delete _delegates[delegateID];
+    _schoolsDelegates[schoolID] = _schoolsDelegates[schoolID].filter(d => d.id !== delegateID);
+  }
+
+  addDelegate(delegate) {
+    _delegates[delegate.id] = delegate;
+    _schoolsDelegates[delegate.school].push(delegate);
+  }
+
+  updateDelegate(delegateID, delta) {
+    ServerAPI.updateDelegate(delegateID, {
+      name: delta.name,
+      email: delta.email,
+    });
+    var delegate = _delegates[delegateID];
+    delegate.name = delta.name;
```

Comment:
This is still a mutative action. Instead, I'd use the spread operator to copy the delegate:
```js
const delegate  = {
  ..._delegates[delegateID],
  name,
  email,
};
_delegates[delegateID] = delegate;
_schoolsDelegates[delegate.school] = /*...*/;
```